src/lib/support/ScopedBuffer.h

```diff
@@ -0,0 +1,152 @@
+namespace Platform {
+
+namespace Impl {
+
+/**
+ * Represents a memory buffer that is auto-freed in the destructor.
+ */
+template <class Impl>
+class ScopedMemoryBufferBase
+{
+public:
+    ScopedMemoryBufferBase() {}
+    ScopedMemoryBufferBase(const ScopedMemoryBufferBase &) = delete;
+    ScopedMemoryBufferBase & operator=(const ScopedMemoryBufferBase & other) = delete;
+
+    ~ScopedMemoryBufferBase() { Free(); }
+
+    void * Ptr() { return mBuffer; }
+    const void * Ptr() const { return mBuffer; }
+
+    template <typename T>
+    inline T * Ptr()
+    {
+        return static_cast<T *>(mBuffer);
+    }
+
+    template <typename T>
+    inline const T * Ptr() const
+    {
+        return static_cast<T *>(mBuffer);
+    }
+
+    ScopedMemoryBufferBase & Alloc(size_t size)
+    {
+        Free();
+        mBuffer = Impl::MemoryAlloc(size);
+        return *this;
+    }
+
+    ScopedMemoryBufferBase & LongTermAlloc(size_t size)
+    {
+        Free();
+        mBuffer = Impl::MemoryAlloc(size, true /* isLongTermAlloc */);
+        return *this;
+    }
+
+    template <typename T>
+    ScopedMemoryBufferBase & Calloc(size_t elementCount)
+    {
+        Free();
+        mBuffer = Impl::MemoryCalloc(elementCount, sizeof(T));
+        return *this;
+    }
+
+    /** Check if a buffer is valid */
+    operator bool() const { return mBuffer != nullptr; }
+
+    /** Take over managing memory from the specified buffer */
+    ScopedMemoryBufferBase & Aquire(void * buffer)
+    {
+        Free();
+        mBuffer = buffer;
+        return *this;
+    }
+
+    /**
+     * Releases the undelying buffer. Buffer stops being managed and will not be
+     * auto-freed.
+     */
+    void * Release()
+    {
+        void * buffer = mBuffer;
+        mBuffer       = nullptr;
+        return buffer;
+    }
+
+    template <typename T>
+    inline T * Release()
+    {
+        void * buffer = mBuffer;
+        mBuffer       = nullptr;
+        return static_cast<T *>(buffer);
+    }
+
+    /** Release memory used */
+    void Free()
+    {
+        if (mBuffer == nullptr)
+        {
+            return;
+        }
+        Impl::MemoryFree(mBuffer);
+        mBuffer = nullptr;
+    }
+
+private:
+    void * mBuffer = nullptr;
+};
+
+} // namespace Impl
+
+/**
+ * Represents a memory buffer allocated using chip::Platform::Memory*Alloc
+ * methods.
+ *
+ * Use for RAII to auto-free after use.
+ */
+class ScopedMemoryBuffer : public Impl::ScopedMemoryBufferBase<ScopedMemoryBuffer>
+{
+    friend class Impl::ScopedMemoryBufferBase<ScopedMemoryBuffer>;
+
+protected:
+    static void MemoryFree(void * p) { chip::Platform::MemoryFree(p); }
+    static void * MemoryAlloc(size_t size) { return chip::Platform::MemoryAlloc(size); }
+    static void * MemoryAlloc(size_t size, bool longTerm) { return chip::Platform::MemoryAlloc(size, longTerm); }
+    static void * MemoryCalloc(size_t num, size_t size) { return chip::Platform::MemoryCalloc(num, size); }
```

Comment:
All of this stuff is inside `namespace chip`, so why the prefixes?

Reply:
Seems we have some include ambiguity. I did not  try to fix it but rather just be explicit. It looks slightly annoying, but not unreadable.